### PR TITLE
Add fetch_yarn_source task

### PR DESCRIPTION
**Pull request by chmeliik**
Roughly equivalent to fetch_npm_source with minor differences in
validation procedure and config file generation.

### CODE REVIEW

**Comment by chmeliik** [comment]
### How to test
Use this branch instead: https://github.com/chmeliik/cachito/tree/fetch-yarn-test (has 2 commits on top of this one, will try to keep it updated)

```shell
curl "localhost:8080/api/v1/requests" \
    -X POST \
    -H 'content-type: application/json' \
    -d '{
          "repo": "https://github.com/chmeliik/cachito-npm-test",
          "ref": "76d6a658f99be99ef6da9455405d874d2d546069",
          "pkg_managers": ["yarn"]
        }'
```

**Comment by chmeliik** [comment]
Shower thoughts: we really need some abstraction for package managers and their tasks or we'll have the same ~200 lines of code for each new package manager.

**Comment by athos-ribeiro** [comment]
> Shower thoughts: we really need some abstraction for package managers and their tasks or we'll have the same ~200 lines of code for each new package manager.

+1

File: cachito/workers/tasks/npm.py [hunk 1]
```diff
@@ -180,13 +170,8 @@ def fetch_npm_source(request_id, package_configs=None):
...
         # The custom CA will be called registry-ca.pem in the "app" directory
-        npm_config_files.append(
-            {
-                "content": base64.b64encode(ca_cert.encode("utf-8")).decode("utf-8"),
-                "path": os.path.join("app", "registry-ca.pem"),
-                "type": "base64",
-            }
-        )
+        ca_path = os.path.join("app", "registry-ca.pem")
+        config_files.append(make_base64_config_file(ca_cert, ca_path))
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/workers/tasks/npm.py:202]: Good catch! I messed up while breaking up one commit into 2. Fixed.

File: cachito/web/models.py [hunk 2]
```diff
@@ -339,7 +339,7 @@ class Dependency(Package):
     This uses the same table as Package, but has different methods.
     """
 
-    pkg_managers_with_dev = ("npm", "pip")
+    pkg_managers_with_dev = ("npm", "pip", "yarn")
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/web/models.py:342]: Oh, I thought `dev` would be a hardcoded `false` for now, but maybe not including it at all is better?

File: cachito/workers/tasks/yarn.py [hunk 3]
```diff
@@ -17,3 +44,169 @@ def cleanup_yarn_request(request_id):
...
+    npm_shrinkwrap = app_source_dir / "npm-shrinkwrap.json"
+    node_modules = app_source_dir / "node_modules"
+
+    def relpath(path):
+        return path.relative_to(root_source_dir)
+
+    if not package_json.is_file():
+        raise ValidationError(
+            f"The {relpath(package_json)} file must be present for the yarn package manager"
+        )
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/workers/tasks/yarn.py:89]: +1, I'll add something to the `tasks.utils` module

File: cachito/web/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (athos-ribeiro) [reply cachito/web/models.py:342]: I came back here just to comment on it :)

I remember we discussed leaving this as false in the past, but then, thinking about it now, that could be just wrong information, right? We should wither set it to a null value or just not include it at all for now.

File: cachito/workers/tasks/utils.py [hunk 4]
```diff
@@ -19,3 +21,79 @@ def make_base64_config_file(content: str, dest_relpath: Union[str, Path]) -> dic
...
+        :param bool expect: expect the file/directory to be present?
+        :param str err_template: error message which may contain {relpath} as a placeholder
+        :raises ValidationError: if the assertion fails
+        """
+        fullpath = self._pkg_dir / path
+
+        if check_presence(fullpath) != expect:
+            relpath = fullpath.relative_to(self._root_dir)
+            err_msg = err_template.format(relpath=relpath)
+            raise ValidationError(f"File check failed for {self._pkg_manager}: {err_msg}")
```
Comment (chmeliik) [inline cachito/workers/tasks/utils.py:99]: Accepting suggestions for a better error message :-)

File: cachito/workers/config.py [hunk 5]
```diff
@@ -270,6 +270,11 @@ def validate_npm_config():
         )
 
 
+# yarn and npm use the same configuration options
+# TODO: rename cachito_nexus_npm_proxy_repo_url?
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/workers/config.py]: Sounds good, added comment in `config.py`

File: cachito/web/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachito/web/models.py]: Yeah not including it sounds better. Fixed

**Comment by chmeliik** [comment]
Addressed comments and reordered commits so that the utils thing for checking file presence comes before the "main" commit

File: cachito/workers/tasks/utils.py (hunk 4 above)
Comment (athos-ribeiro) [reply cachito/workers/tasks/utils.py:99]: This LGTM

File: cachito/workers/tasks/yarn.py [hunk 6]
```diff
@@ -17,3 +43,131 @@ def cleanup_yarn_request(request_id):
...
+    - npm-shrinkwrap.json must not be present
+    - node_modules/ must not be present
+
+    :param RequestBundleDir bundle_dir: the ``RequestBundleDir`` object for the request
+    :param list[str] subpaths: a list of subpaths in the source repository of yarn packages
+    :raises CachitoError: if the repository is missing the required files or contains invalid
+        files/directories
+    """
+    for subpath in subpaths:
+        assert_files = AssertPackageFiles("yarn", bundle_dir.source_root_dir, package_path=subpath)
```
Comment (athos-ribeiro) [inline cachito/workers/tasks/yarn.py:65]: Nice!

**Review by athos-ribeiro — APPROVED** [review]

**Pull request merged**